File: xdm-core/src/main/java/xdm/core/downloaders/SpeedLimiter.java
//package xdm.core.downloaders;
//
//import java.util.concurrent.CountDownLatch;
//import java.util.concurrent.TimeUnit;
//import org.slf4j.Logger;
//import org.slf4j.LoggerFactory;
//import xdm.core.Config;
//
//public class SpeedLimiter {
//  private long lastTick, lastBytes;
//  private long lastChecked = System.currentTimeMillis();
//  private int cachedSpeedLimit = -2;
//  private CountDownLatch sleepHandle = new CountDownLatch(1);
//  private AbstractDownloader downloader;
//  private final Logger logger = LoggerFactory.getLogger(SpeedLimiter.class);
//
//  public SpeedLimiter(AbstractDownloader downloader) {
//    this.downloader = downloader;
//  }
//
//  private int getCachedSpeedLimit() {
//    long now = System.currentTimeMillis();
//    if (now - lastChecked > 3000 || cachedSpeedLimit == -2) {
//      lastChecked = now;
//      cachedSpeedLimit = getGlobalSpeedLimit();
//    }
//    return cachedSpeedLimit;
//  }
//
//  private int getGlobalSpeedLimit() {
//    int speedLimit = 0;
//    Config config = Config.getInstance();
//    synchronized (config) {
//      if (config.getSpeedLimit() > 0) {
//        speedLimit = config.getSpeedLimit();
//      }
//    }
//    return speedLimit;
//  }
//
//  public void throttleIfNeeded() {
//    int speedLimit = getCachedSpeedLimit();
//    if (speedLimit < 1) return;
//    if (this.downloader == null) {
//      return;
//    }
//    if (lastBytes == 0 || lastTick == 0) {
//      lastBytes = downloader.getDownloaded();
//      lastTick = System.currentTimeMillis();
//      return;
//    }
//    synchronized (this.downloader) {
//      double maxBytesPerMS = (double) speedLimit * 1024 / 1000;
//      long now = System.currentTimeMillis();
//      long actualTimeSpent = now - lastTick;
//      if (actualTimeSpent < 1) return;
//      long bytes = downloader.getDownloaded();
//      long diff = bytes - lastBytes;
//      lastBytes = bytes;
//      lastTick = now;
//      double expectedTimeSpent = maxBytesPerMS > 0 ? diff / maxBytesPerMS : 0;
//      if (actualTimeSpent < expectedTimeSpent) {
//        try {
//          sleep((int) Math.ceil(expectedTimeSpent - actualTimeSpent));
//        } catch (Exception ex) {
//          logger.info("Exception while throttling");
//        }
//      }
//    }
//  }
//
//  private void sleep(int interval) {
//    try {
//      sleepHandle.await(interval, TimeUnit.MILLISECONDS);
//    } catch (Exception ex) {
//      // Swallow
//      ex.printStackTrace();
//    }
//  }
//
//  public void wakeIfSleeping() {
//    sleepHandle.countDown();
//  }
//}

File: app/XDM/XDM.Core/Downloader/SpeedLimiter.cs
using System;
using System.Threading;
using TraceLog;
using XDM.Core.Util;

namespace XDM.Core.Downloader
{
    public class SpeedLimiter
    {
        private long lastTick, lastBytes;
        private ManualResetEvent sleepHandle = new ManualResetEvent(false);
        private long lastChecked = Helpers.TickCount();
        private int cachedSpeedLimit = -2;

        public int SpeedLimit => cachedSpeedLimit;

        public void WakeIfSleeping()
        {
            this.sleepHandle.Set();
        }

        private int GetCachedSpeedLimit()
        {
            var now = Helpers.TickCount();
            if (now - lastChecked > 3000 || cachedSpeedLimit == -2)
            {
                lastChecked = now;
                cachedSpeedLimit = GetGlobalSpeedLimit();
            }
            return cachedSpeedLimit;
        }

        private int GetGlobalSpeedLimit()
        {
            int speedLimit = 0;
            lock (Config.Instance)
            {
                if (Config.Instance.EnableSpeedLimit && Config.Instance.DefaltDownloadSpeed > 0)
                {
                    speedLimit = Config.Instance.DefaltDownloadSpeed;
                }
            }
            return speedLimit;
        }

        public void ThrottleIfNeeded(IBaseDownloader downloader)
        {
            int speedLimit = GetCachedSpeedLimit();
            if (speedLimit < 1) return;
            if (lastBytes == 0 || lastTick == 0)
            {
                lastBytes = downloader.GetDownloaded();
                lastTick = Helpers.TickCount();
                return;
            }
            try
            {
                downloader.Lock.EnterWriteLock();
                var maxBytesPerMS = (double)speedLimit * 1024 / 1000;
                var now = Helpers.TickCount();
                var actualTimeSpent = now - lastTick;
                if (actualTimeSpent < 1) return;
                var bytes = downloader.GetDownloaded();
                var diff = bytes - lastBytes;
                lastBytes = bytes;
                lastTick = now;
                var expectedTimeSpent = diff / maxBytesPerMS;

                if (actualTimeSpent < expectedTimeSpent)
                {
                    try
                    {
                        sleep((int)Math.Ceiling(expectedTimeSpent - actualTimeSpent));
                    }
                    catch (Exception ex) { Log.Debug(ex, "Exception while throttling"); }
                }
            }
            finally
            {
                downloader.Lock.ExitWriteLock();
            }
        }

        private void sleep(int interval)
        {
            sleepHandle.WaitOne(interval);
        }
    }
}
